### ISSUE TITLE
Feature Request: enable rechanging tf.device of a tensor

### ISSUE THREAD

**Issue by bar513** [tensorflow/tensorflow#15341]
It seems that once a tensor's GPU was defined using tf.device, the GPU cannot be changed anymore. 
When loading saved graphs, the graph will use the same GPU that was chosen years ago and it cannot be set again.

thanks.

**Comment by angersson** [comment]
Can you provide an example that demonstrates this?

**Comment by bar513** [comment]
Here I demonstrate the missing feature on the mnist classic example

```
# using tensorflow 1.3 

from tensorflow.examples.tutorials.mnist import input_data
import tensorflow as tf


# create a model and save the graph and the tensors for later training
def createModelAndSave(my_gpu):
    learning_rate = 0.01
    with tf.device(my_gpu): # all model's tensors are on this gpu, for example '/gpu:0'
        x = tf.placeholder("float", [None, 784])
        y = tf.placeholder("float", [None, 10])

        # Set model weights
        W = tf.Variable(tf.zeros([784, 10]))
        b = tf.Variable(tf.zeros([10]))

        model = tf.nn.softmax(tf.matmul(x, W) + b)

        # cost function
        cost_function = -tf.reduce_sum(y*tf.log(model))

        # Gradient descent
        optimizer = tf.train.GradientDescentOptimizer(learning_rate).minimize(cost_function)

    # save all tensors in the collection
    tf.add_to_collection('cost_function',cost_function)
    tf.add_to_collection('optimizer',optimizer)
    tf.add_to_collection('x', x)
    tf.add_to_collection('y', y)

    saver = tf.train.Saver()
    with tf.Session() as sess:
        init = tf.global_variables_initializer()
        sess.run(init)
        saver.save(sess,'myModel.ckpt')


# load the saved graph and the tensors for training
def loadModelAndTrain():
    mnist = input_data.read_data_sets("MNIST_data", one_hot=True)
    training_iteration = 1000
    batch_size = 100

    # load graph
    tf.train.import_meta_graph('myModel.ckpt.meta')
    with tf.Session() as sess:
        # load tensors
        cost_function = sess.graph.get_collection("cost_function")[0]
        optimizer = sess.graph.get_collection("optimizer")[0]
        x = sess.graph.get_collection("x")[0]
        y = sess.graph.get_collection("y")[0]

        # ~~~~~~~
        # At this point the tensors and the graph are loaded but all of them are on GPU that was chosen when the model was saved.
        # for example, the training will run on '/gpu:0' as chosen years ago.
        # assuming we don't have the function to create the model again, we cannot relocate the tensors to anther GPU.
        # ~~~~~~~
        init = tf.global_variables_initializer()
        sess.run(init)

        # Training cycle - all on the chosen GPU
        for iteration in range(training_iteration):
            avg_cost = 0.
            total_batch = int(mnist.train.num_examples/batch_size)
            for i in range(total_batch):
                batch_xs, batch_ys = mnist.train.next_batch(batch_size)
                sess.run(optimizer, feed_dict={x: batch_xs, y: batch_ys})
                avg_cost += sess.run(cost_function, feed_dict={x: batch_xs, y: batch_ys})/total_batch
            # Display logs per iteration step
            print ("Iteration:", '%04d' % (iteration + 1), "cost=", "{:.9f}".format(avg_cost))


if __name__ == "__main__":

    # ~~~~~ first run createModelAndSave(), later comment this function and run loadModelAndTrain()

    createModelAndSave('/gpu:0') # save a model to a specific GPU
    # loadModelAndTrain() # load the saved model and train it, cannot rechoose GPU

```

**Comment by ppwwyyxx** [comment]
https://www.tensorflow.org/api_docs/python/tf/train/import_meta_graph


import_meta_graph(
    meta_graph_or_file,
    __clear_devices=False,__
    import_scope=None,
    **kwargs
)

**Comment by bar513** [comment]
Thank you for your answer.
It's indeed clear the devices and rest them to their default GPU:0, but it is still not possible to choose another one. (GPU:2 for example).
we are a team that work  on several GPUs. The feature to LOAD each model to a different GPU is very important and still missing.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
The original poster has replied to this issue after the stat:awaiting response label was applied.

**Comment by mrry** [comment]
In addition to setting `clear_devices=True`, you can wrap the call to `import_meta_graph()` in another `with tf.device(...):`  block, and either specify a device name to be assigned to all nodes, or a function (taking a `tf.Operation` and returning a device name) that will be called once for each operation in the imported graph.
